File: android-core/src/main/java/com/mparticle/internal/CoreCallbacks.java
package com.mparticle.internal;

import android.app.Activity;
import android.net.Uri;

import com.mparticle.MParticleOptions;

import org.json.JSONArray;

import java.lang.ref.WeakReference;
import java.util.Map;

public interface CoreCallbacks {
    boolean isBackgrounded();
    int getUserBucket();
    boolean isEnabled();
    void setIntegrationAttributes(int kitId, Map<String, String> integrationAttributes);
    Map<String, String> getIntegrationAttributes(int kitId);
    WeakReference<Activity> getCurrentActivity();
    JSONArray getLatestKitConfiguration();
    MParticleOptions.DataplanOptions getDataplanOptions();
    boolean isPushEnabled();
    String getPushSenderId();
    String getPushInstanceId();
    Uri getLaunchUri();
    String getLaunchAction();
    void replayAndDisableQueue();
    KitListener getKitListener();

    interface KitListener {

        void kitFound(int kitId);
        void kitConfigReceived(int kitId, String configuration);
        void kitExcluded(int kitId, String reason);
        void kitStarted(int kitId);
        void onKitApiCalled(int kitId, Boolean used, Object... objects);
        void onKitApiCalled(String methodName, int kitId, Boolean used, Object... objects);

        KitListener EMPTY = new KitListener() {
            public void kitFound(int kitId) {}
            public void kitConfigReceived(int kitId, String configuration) { }
            public void kitExcluded(int kitId, String reason) { }
            public void kitStarted(int kitId) { }
            public void onKitApiCalled(int kitId, Boolean used, Object... objects) { }
            public void onKitApiCalled(String methodName, int kitId, Boolean used, Object... objects) { }
        };
    }
}

File: android-core/src/main/kotlin/com/mparticle/internal/CoreCallbacks.kt
package com.mparticle.internal

import android.app.Activity
import android.net.Uri
import androidx.annotation.WorkerThread
import com.mparticle.MParticleOptions.DataplanOptions
import org.json.JSONArray
import java.lang.ref.WeakReference

interface CoreCallbacks {
    fun isBackgrounded(): Boolean

    fun getUserBucket(): Int

    fun isEnabled(): Boolean

    fun setIntegrationAttributes(kitId: Int, integrationAttributes: Map<String, String>)

    fun getIntegrationAttributes(kitId: Int): Map<String, String>?

    fun getCurrentActivity(): WeakReference<Activity>?

    @WorkerThread
    fun getLatestKitConfiguration(): JSONArray?

    fun getDataplanOptions(): DataplanOptions?

    fun isPushEnabled(): Boolean

    fun getPushSenderId(): String?

    fun getPushInstanceId(): String?

    fun getLaunchUri(): Uri?

    fun getLaunchAction(): String?

    fun getKitListener(): KitListener?

    interface KitListener {
        fun kitFound(kitId: Int)

        fun kitConfigReceived(kitId: Int, configuration: String?)

        fun kitExcluded(kitId: Int, reason: String?)

        fun kitStarted(kitId: Int)
        fun onKitApiCalled(kitId: Int, used: Boolean?, vararg objects: Any?)
        fun onKitApiCalled(methodName: String?, kitId: Int, used: Boolean?, vararg objects: Any?)

        companion object {
            @JvmField
            val EMPTY: KitListener = object : KitListener {
                override fun kitFound(kitId: Int) {}
                override fun kitConfigReceived(kitId: Int, configuration: String?) {}
                override fun kitExcluded(kitId: Int, reason: String?) {}
                override fun kitStarted(kitId: Int) {}
                override fun onKitApiCalled(kitId: Int, used: Boolean?, vararg objects: Any?) {}
                override fun onKitApiCalled(methodName: String?, kitId: Int, used: Boolean?, vararg objects: Any?) {}
            }
        }
    }
}